v2/lints/mozilla/lint_mp_pss_parameters_encoding_correct.go

```diff
@@ -0,0 +1,98 @@
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/************************************************
+
+https://www.mozilla.org/en-US/about/governance/policies/security-group/certs/policy/
+
+Section 5.1.1 RSA
+
+RSASSA-PSS with SHA-256, MGF-1 with SHA-256, and a salt length of 32 bytes.
+
+The encoded AlgorithmIdentifier MUST match the following hex-encoded bytes:
+
+304106092a864886f70d01010a3034a00f300d0609608648016503040201
+0500a11c301a06092a864886f70d010108300d0609608648016503040201
+0500a203020120
+
+RSASSA-PSS with SHA-384, MGF-1 with SHA-384, and a salt length of 48 bytes.
+
+The encoded AlgorithmIdentifier MUST match the following hex-encoded bytes:
+
+304106092a864886f70d01010a3034a00f300d0609608648016503040202
+0500a11c301a06092a864886f70d010108300d0609608648016503040202
+0500a203020130
+
+RSASSA-PSS with SHA-512, MGF-1 with SHA-512, and a salt length of 64 bytes.
+
+The encoded AlgorithmIdentifier MUST match the following hex-encoded bytes:
+
+304106092a864886f70d01010a3034a00f300d0609608648016503040203
+0500a11c301a06092a864886f70d010108300d0609608648016503040203
+0500a203020140
+************************************************/
+
+import (
+	"bytes"
+	"encoding/hex"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v2/lint"
+	"github.com/zmap/zlint/v2/util"
+)
+
+type rsaPssAidEncoding struct{}
+
+var allowedAidEncodingsForPSS = [3]string{
+	"304106092a864886f70d01010a3034a00f300d06096086480165030402010500a11c301a06092a864886f70d010108300d06096086480165030402010500a203020120",
+	"304106092a864886f70d01010a3034a00f300d06096086480165030402020500a11c301a06092a864886f70d010108300d06096086480165030402020500a203020130",
+	"304106092a864886f70d01010a3034a00f300d06096086480165030402030500a11c301a06092a864886f70d010108300d06096086480165030402030500a203020140"}
+
+func (l *rsaPssAidEncoding) Initialize() error {
+	return nil
+}
+
+func (l *rsaPssAidEncoding) CheckApplies(c *x509.Certificate) bool {
+	return c.SignatureAlgorithmOID.Equal(util.OidRSASSAPSS)
+}
+
+func (l *rsaPssAidEncoding) Execute(c *x509.Certificate) *lint.LintResult {
+	signatureAlgoID, err := util.GetSignatureAlgorithmInTBSEncoded(c)
+
+	if err != nil {
+		return &lint.LintResult{Status: lint.Error, Details: "error reading signatureAlgorithm from TBS"}
+	}
+
+	for _, encoding := range allowedAidEncodingsForPSS {
+		expectedEncoding, _ := hex.DecodeString(encoding)
+		if bytes.Equal(signatureAlgoID, expectedEncoding) {
+			return &lint.LintResult{Status: lint.Pass}
+		}
+	}
```

Comment:
I think it would be better to use the approach taken in https://github.com/zmap/zlint/blob/5b2df5c915f1d73dbfcd4f2ac2a6241103fbc4a3/v2/util/algorithm_identifier.go#L36 and https://github.com/zmap/zlint/blob/5b2df5c915f1d73dbfcd4f2ac2a6241103fbc4a3/v2/lints/rfc/lint_tbs_signature_rsa_encryption_parameter_not_null.go#L43

E.g. extract the signature OID from `RawSubjectPublicKeyInfo` with `cryptobyte`, and then use that as a lookup into a map of expected `[]byte`.

Reply:
See also https://github.com/zmap/zlint/pull/377#discussion_r386237154. In the case of RSAPSS is not possible to extract an OID and check that it maps to the correct value. I believe to check whether the encoding is one of the three allowed ones is (in the case of RSAPSS) better. Ok?

Reply:
Makes sense thanks! I replied in the other comment thread.